src/Control/Email/Email.php

```
@@ -769,237 +399,150 @@ public function getHTMLTemplate()
+        }
+        if ($htmlRender) {
+            $this->html($htmlRender);
+        }
+
+        $this->dataAtLastRender = $this->getData();
     }
 
-    /**
-     * @deprecated 4.12.0 Will be removed without equivalent functionality to replace it
-     *
-     * @return bool
-     */
-    public function hasPlainPart()
+    private function getDataHasChangedSinceLastRender(): bool
     {
-        Deprecation::notice('4.12.0', 'Will be removed without equivalent functionality to replace it');
-        if ($this->getSwiftMessage()->getContentType() === 'text/plain') {
-            return true;
-        }
-        return (bool) $this->findPlainPart();
+        return $this->getSerializedData($this->getData()) !== $this->getSerializedData($this->dataAtLastRender);
     }
 
     /**
-     * @deprecated 4.12.0 Will be removed without equivalent functionality to replace it
-     *
-     * Automatically adds a plain part to the email generated from the current Body
-     *
-     * @return $this
+     * Quickly serialize viewable data while removing any non-seriliazable properties such as closures
      */
-    public function generatePlainPartFromBody()
+    private function getSerializedData(ViewableData $viewableData): string
     {
-        Deprecation::notice('4.12.0', 'Will be removed without equivalent functionality to replace it');
-        $plainPart = $this->findPlainPart();
-        if ($plainPart) {
-            $this->getSwiftMessage()->detach($plainPart);
-        }
-        unset($plainPart);
-
-        $this->getSwiftMessage()->addPart(
-            Convert::xml2raw($this->getBody()),
-            'text/plain',
-            'utf-8'
-        );
-
-        return $this;
+        $data = $viewableData instanceof ArrayData ? (object) $viewableData->toMap() : $viewableData;
+        return serialize(array_map(function ($property) {
+            try {
+                return serialize($property);
+            } catch (Exception) {
```

This is too broad a brush to paint with. If this is explicitly to catch the "Serialization of 'Closure' is not allowed" exception, then we should check for that message (being relaxed enough to have "Serialization of '[a-zA-Z]+' is not allowed") and throw everything else - or find a different way of handling this (e.g. check the type of `$property` and just not serialise types that can't be serialized - though based on the [note in the docs](https://www.php.net/manual/en/function.serialize.php#refsect1-function.serialize-notes) that would be hard to do reliably).

Otherwise we risk ignoring exceptions caused by some code in a `_serialize()` method, for example.

---

Code has been removed